### ISSUE TITLE
Memory leak: autograd contexts not being garbage collected when assigned output Tensors as attributes

### ISSUE THREAD

**Issue by patrick-kidger** [pytorch/pytorch#25340]
## 🐛 Bug

Autograd contexts (I think they're called this - the `ctx` variable used with `autograd.Function`) are not garbage collected by Python when the *output* of the `autograd.Function` is assigned to them as an attribute. In particular this means that every other attribute of the context also sticks around, and memory usage quickly spirals out of control.

MWE:
```python
import gc
import time
import torch
import weakref


class Test(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, refs):
        refs.append(weakref.ref(ctx))
        y = x.clone()
        ctx.y = y
        return y


if __name__ == '__main__':
    refs = []  # weakrefs to ctx objects
    x = torch.rand(8, 8, requires_grad=True)
    y = Test.apply(x, refs)
    print(refs)
    del y
    gc.collect()
    # Another bug mentions sleeping somehow helping. It doesn't
    # help here.
    time.sleep(20)
    # the weakref should died by now
    print(refs)  # Bug! Weakref is still alive
```

We expect the weakref to die because the context object should have been garbage collected. This does not occur.

Some observations: deleting the `ctx.y = y` line or changing it to `ctx.y = x` gives the expected behaviour. Removing the `requires_grad=True` from `x` also gives the expected behaviour.

I had a look through existing issues and found [this](https://github.com/pytorch/pytorch/issues/18643), but I'm not sure if they're actually related.

 - PyTorch Version (e.g., 1.0): 1.2
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.7.4

cc @ezyang @SsnL @albanD @zou3519

**Comment by patrick-kidger** [comment]
Actually I think the cause is pretty obvious: `y` has a reference to `ctx` (via its `grad_fn` attribute), and `ctx` obviously has a reference to `y`, giving a reference cycle.

I'm guessing the `grad_fn` attribute is handled in the C++ side so the cycle isn't being correctly detected and collected by Python's GC.

**Comment by patrick-kidger** [comment]
A little further testing: using `save_for_backward` in place of assigning as an attribute also gives expected results.

I'm aware `save_for_backward` is the preferred method for storing Tensors. I stumbled across this bug in a somewhat more involved context, when the thing being assigned isn't actually a Tensor itself, but is a complicated structure that happens to include a reference to the output Tensor.

**Comment by patrick-kidger** [comment]
With apologies for leaving 4 comments in a row -

For anyone else reading this who's hit the same bug (and cannot for whatever reason use `save_for_backward`), the easiest way around it that I've found is to break the reference cycle by instead assigning `ctx.y = y.detach()`. In particular *not* breaking the cycle by doing `return y.detach()`: this will successfully break the reference cycle and solve this issue, but then runs afoul of [another bug](https://github.com/pytorch/pytorch/issues/24413).

**Comment by soumith** [comment]
https://github.com/pytorch/pytorch/issues/3818#issuecomment-346236024 and https://github.com/pytorch/pytorch/issues/18643 look suspiciously close.

cc: @ezyang @albanD I think this has come up before, though I might just have lapsed memory.

**Comment by soumith** [comment]
@ezyang please bump priority if needed.

**Comment by ezyang** [comment]
Yes this is exactly #3818. We have a patch from @skrah that is very close to being done; I invite you to try the PR at #24368 to see if it works for you (we have some ancillary correctness concerns with the PR which is why it hasn't merged yet, but it should fix the leak you see here.)

**Comment by albanD** [comment]
@ezyang I think this one is slightly different no? It is the simple case where you save an output in the ctx. That is why `save_for_backward` exists. There is no backward involved in the example above.

**Comment by patrick-kidger** [comment]
Judging by [this comment on #3818](https://github.com/pytorch/pytorch/issues/3818#issuecomment-346236024), this seems to be a different reference cycle.

I think the general reference cycle here is essentially
`output --[.grad_fn]--> ctx --[.some_datastructure]--> arbitrary_datastructure --[.some_attribute]--> output`, if you follow my notation.

Since there's no control over what can be set as an attribute on `ctx`, I'm guessing the only generic way out of this would be to `.detach()` every Tensor that is outputted from the forward (and backward?) calls to `autograd.Function`. (And contrary to what I said in my last comment, this approach doesn't actually seem to run afoul of the other bug I mentioned. I don't know enough about the PyTorch internals to speculate why that should be.)

**Comment by albanD** [comment]
Two solutions:
- If you don't use double backward (gradients of gradients), then detaching the Tensor that you save will do the trick.
- If you do need double backward, then you cannot use detach (or the gradients of gradients can be wrong) and you will have to make sure that you **never** save a Tensor that is either an input or an output inside ctx. You will need to save these with save_for_backward and then put them back in your data structure at the beginning of the backward method.

Unfortunately, there is not much we can do here as we don't want to traverse every attribute of ctx and data structures in there to try and find an output.

Feel free to reopen this issue if you encounter more problems working around this.
